src/libraries/System.Net.Http/src/System/Net/Http/Headers/HeaderDescriptor.cs

```diff
@@ -169,7 +186,18 @@ public string GetHeaderValue(ReadOnlySpan<byte> headerValue, Encoding? valueEnco
                 }
             }
 
-            return (valueEncoding ?? HttpRuleParser.DefaultHttpEncoding).GetString(headerValue);
+            Encoding encoding = valueEncoding ?? HttpRuleParser.DefaultHttpEncoding;
+            replaceDangerousCharacters = replaceDangerousCharacters && headerValue.IndexOfAny(s_dangerousCharacterBytes) >= 0;
+            int length = encoding.GetCharCount(headerValue);
+            return string.Create(length, new CreateHeaderStringState(headerValue, encoding, replaceDangerousCharacters), static (chars, s) =>
+            {
+                int doubleCheck = s.Encoding.GetChars(s.Bytes, chars);
+                Debug.Assert(chars.Length == doubleCheck);
+                if (s.ReplaceDangerousCharacters)
+                {
+                    chars.ReplaceAny(s_dangerousCharacters, ' ');
+                }
+            });
```

Comment:
The case where we find any such character should be a super rare.
It'd probably be fine to simplify this to something like
```c#
string value = (valueEncoding ?? HttpRuleParser.DefaultHttpEncoding).GetString(headerValue);

if (replaceDangerousCharacters && value.ContainsAny(s_dangerousCharacters))
{
    value = value.Replace('\r', ' ').Replace('\n', ' ').Replace('\0', ' ');
}

return value;
```

Reply:
The expected encodings for the 99.9...% case here are Latin1/UTF8/ASCII. With those it wouldn't matter since those bytes map into the same characters. #116859 tracks documenting that we really don't recommend using anything else.

So it should be extremely rare, but fair point that we don't have to break them here.

My previous suggestion was something like
```c#
string value = (valueEncoding ?? HttpRuleParser.DefaultHttpEncoding).GetString(headerValue);

if (replaceDangerousCharacters && value.ContainsAny(s_dangerousCharacters))
{
    value = value.Replace('\r', ' ').Replace('\n', ' ').Replace('\0', ' ');
}

return value;
```